### PR TITLE
DR-786 Add support for partitioned tables.

### PR DESCRIPTION
**Pull request by danxmoran**
This enables partitioning at the dataset layer, but doesn't make use of it at the snapshot layer.

### CODE REVIEW

**Comment by danxmoran** [comment]
Test failure was another IAM blip, will hold off on the retry-tests-and-pray cycle until I see if I need to push new code.

File: src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java
Comment (thathert) [inline src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java]: The `flatMap` is throwing me off -- the name makes me think it is applying a function over the sequence of characters in the string. Can we keep the `stream -> filter -> findFirst`  and find an alternative? I think it will help the readability here.

File: src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java
Comment (thathert) [inline src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java]: What other types are there? Worth considering exposing this as an option to stewards now or in the future?

File: src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java
Comment (thathert) [inline src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java]: I assume `null` here means use ingest time? Can you add a comment here?

File: src/main/resources/data-repository-openapi.yaml
Comment (thathert) [inline src/main/resources/data-repository-openapi.yaml]: Our convention is to use camelCase inside of the openapi yaml (except for the ga4gh endpoints and apparently we left it in for `created_date` in the `sort` enums..

File: src/test/java/bio/terra/service/tabulardata/google/BigQueryPdaoTest.java
Comment (thathert) [inline src/test/java/bio/terra/service/tabulardata/google/BigQueryPdaoTest.java:290]: Can you add a comment about what this `$__PARTITIONS_SUMMARY__` thing is?

**Review by thathert — APPROVED** [review]
A few readability things and a request around conventions in the openapi yaml, but otherwise this looks great. Thanks!

File: src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java]: The functional-style APIs are one of the main benefits of `Optional` over `null`. We could translate the `flatMap` to a ternary:
```
optional.isPresent() ? someMethod(optional.get()) : Optional.empty()
```
But IntelliJ flags it as a lint warning, and suggests using the functional APIs instead.

I'll rewrite to use `null`s and the ternary.

File: src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java]: There are no other types (yet?). I'll add a comment.

File: src/main/resources/data-repository-openapi.yaml
Comment (danxmoran) [reply src/main/resources/data-repository-openapi.yaml]: There are a few other spots (`array_of` is one I can think of off the top of my head). I'll update these, should I make a ticket for correcting the other inconsistencies?

File: src/main/java/bio/terra/service/dataset/DatasetJsonConversion.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/dataset/DatasetJsonConversion.java]: I think Optional should only be used for output parameters. Using it just so you can do its streaming function seems to me to be a way to obscure the logic. How about:
```suggestion
if (tableModel.getPartitionColumn() != null) {
    datasetTable.partitionColumn(columnMap.get(tableModel.getPartitionColumn());
}
```

File: src/main/resources/data-repository-openapi.yaml
Comment (ddietterich) [inline src/main/resources/data-repository-openapi.yaml]: Instead of putting partitionColumn here, I suggest breaking out separate structures for each of the strategy types.
Then you make it impossible to express things like: supplying a partition column when it isn't allowed for that strategy.

See how Rori did that for the snapshot interface. I think it is much cleaner. Also, in OpenAPI V3, we can restrict so only one choice can be valid. With V2, we have to validate that the other choices are null.

File: src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/dataset/DatasetRequestValidator.java]: As mentioned on the OpenAPI, this would be simpler if you broke out the input into separate objects.

File: src/main/java/bio/terra/service/dataset/DatasetTable.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/dataset/DatasetTable.java]: Another alternative would be to make a base class for partitionStrategy and build the right one.
You only need the details inside PDAO, I think.

File: src/main/java/bio/terra/service/dataset/DatasetTableDao.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/dataset/DatasetTableDao.java]: Another unneeded use of Optional.

File: src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java]: If we made a class structure for the partition strategies, we could have them self-generate, instead of using this logic.

File: src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java]: Should we always have a partitionStrategy; e.g., NONE. Even if it isn't in the database, it would allow a switch across types or a class structure instead of this logic flow.

File: src/main/java/bio/terra/service/tabulardata/google/BigQueryProject.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/tabulardata/google/BigQueryProject.java]: Again, this smells like a good case for a subclass structure with a create method.

**Review by ddietterich — CHANGES_REQUESTED** [review]

**Comment by ddietterich** [comment]
I think the structure in the OpenAPI and the PDAO could be improved by separating out classes, and for the dao and pdao, a base class and subclasses to represent the different kinds of partitioning and the code generation they do. I think it would make the code more readable and more easily extensible in the future.

File: src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java]: The tricky part here is that `TimePartitioning` and `RangePartitioning` don't share a superclass themselves, so it's tough to abstract over a method that returns one or the other. We could have an interface like:
```java
interface PartitionMode {
  void setPartition(TableDefinition tableDef)
}
```
And then the internals could no-op or call the appropriate setter. That'd blur the lines pretty heavily between DAO and PDAO lands, though.

Comment (danxmoran) [reply src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java]: I guess this would work:
```java
interface PartitionMode {
  TimePartitioning getTimePartitioning()
  RangePartitioning getRangePartitioning()
}
```
And the different subclasses return `null` for the method(s) that they don't know about.

File: src/main/java/bio/terra/service/dataset/DatePartitionMode.java
Comment (danxmoran) [inline src/main/java/bio/terra/service/dataset/DatePartitionMode.java]: Note to self: Actually add these Javadocs, here and elsewhere.

**Comment by danxmoran** [comment]
@ddietterich I've pushed an implementation based on your suggested approach. It does feel cleaner, though it kinda blurs the lines between the DAO / PDAO layers (not sure if we care at this point). Can you take a look?

**Comment by danxmoran** [comment]
@thathert the code looks pretty different from when you approved it, do you want to look again?

File: src/main/java/bio/terra/service/dataset/DatasetTableDao.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/dataset/DatasetTableDao.java]: I was thinking things like this could be in the subclasses. Could have the base class implement the null return... but it would require an object for the no mode case.

Don't feel strongly about it.

**Review by ddietterich — APPROVED** [review]
LGTM
One small comment.
And yes, it does mix the DAO and PDAO. But we are exposing a BQ-specific thing with an odd API.

File: src/main/java/bio/terra/service/dataset/DatasetTableDao.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/dataset/DatasetTableDao.java]: This is where I could see using a `json` column to store the partitioning options could make sense, so there's just one method to override that returns whatever-it-needs-to, with a complementary `static` builder to reconstruct the appropriate mode based on the JSON.

Comment (danxmoran) [reply src/main/java/bio/terra/service/dataset/DatasetTableDao.java]: Actually, that might help with the DAO/PDAO blurring too... I'm going to play with it a bit before merging this.

**Review by thathert — APPROVED** [review]
LGTM -- interested to see what you find about the one open comment in `DatasetTableDao`

File: src/main/java/bio/terra/service/dataset/DatasetJsonConversion.java
Comment (danxmoran) [inline src/main/java/bio/terra/service/dataset/DatasetJsonConversion.java:131]: Note to self: Should update this method too, so partitioning settings flow back the other way?

File: src/main/java/bio/terra/service/dataset/DatasetDao.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/dataset/DatasetDao.java]: Minor: How come the switch to Exception? I'd prefer a more precise list.

File: src/main/java/bio/terra/service/dataset/DatasetTableDao.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/dataset/DatasetTableDao.java]: The norm for our code is to list all `@Autowired` in the constructor and assign. That is what is happening right below this autowire.

File: src/main/resources/db/changesets/20200312_addpartitions.yaml
Comment (ddietterich) [inline src/main/resources/db/changesets/20200312_addpartitions.yaml]: suggest jsonb as noted above

File: src/main/java/bio/terra/service/dataset/DatasetTableDao.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/dataset/DatasetTableDao.java:127]: We need to understand how we version this. If we need to add/change the json schema, how do we do it?

One idea would be:
1. Store this in a jsonb column (you have it as json), so it is queryable
2. Add a version string as a member of the BigQueryPartitionConfig class
3. To deserialize, read the version out of the json and read the json as a string
4. Based on the version, use objectMapper to map to the matching class.




serialize by retrieving the version

File: src/main/resources/db/changesets/20200312_addpartitions.yaml
Comment (ddietterich) [inline src/main/resources/db/changesets/20200312_addpartitions.yaml]: Need to apply the versioning scheme here also

**Review by ddietterich — CHANGES_REQUESTED** [review]
This is excellent. I really, really like encapsulating everything in the BQ-specific class.

My biggest ask is to do something to handle versioning of the json object in the database. I have a comment on that.

I'd like to understand why you switched a bunch of methods from specific exceptions to Exception. I'd prefer the finer grained approach.

File: src/main/java/bio/terra/service/dataset/DatasetDao.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/dataset/DatasetDao.java]: Just to make it compile, so I could run tests. Will switch them back.

File: src/main/java/bio/terra/service/dataset/DatasetTableDao.java
Comment (danxmoran) [inline src/main/java/bio/terra/service/dataset/DatasetTableDao.java:128]: This check caught that the object mapper wasn't actually persisting all the `private` members (eek!). Added a case to the `DatasetDaoTest` to guard against regressions there.

**Review by thathert — APPROVED** [review]
Looks good, I like the JSON approach. Thanks for switching to JSONB and adding versioning.

File: src/main/resources/db/changesets/20200312_addpartitions.yaml
Comment (fboulnois) [inline src/main/resources/db/changesets/20200312_addpartitions.yaml:11]: +1!

File: src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetMetadataStep.java
Comment (fboulnois) [inline src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetMetadataStep.java:51]: Was this change also just for testing? You could `catch` other errors separately, but this change seems reasonable to me either way.

**Review by fboulnois — APPROVED** [review]
Using `jsonb` will be really useful. I have one other question:

File: src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetMetadataStep.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetMetadataStep.java:51]: The existing exception message wasn't SQL-specific, so I thought it'd be reasonable to widen the `catch` here. I could go the opposite direction and catch each type individually, with tailored messages in the rethrown exceptions, if it'd be useful.

**Review by ddietterich — APPROVED** [review]
LGTM

**Pull request merged**
